pkg/server/delta/v3/server.go

```diff
@@ -44,6 +132,340 @@ func NewServer(ctx context.Context, config cache.ConfigWatcher, callbacks Callba
 	}
 }
 
+func (s *server) processDelta(str stream.DeltaStream, reqCh <-chan *discovery.DeltaDiscoveryRequest, defaultTypeURL string) error {
+	// increment stream count
+	streamID := atomic.AddInt64(&s.streamCount, 1)
+
+	// unique nonce generator for req-resp pairs per xDS stream; the server
+	// ignores stale nonces. nonce is only modified within send() function.
+	var streamNonce int64
+
+	// a collection of stack alloceated watches per request type
+	var values watches
+	values.Init()
+
+	defer func() {
+		values.Cancel()
+		if s.callbacks != nil {
+			s.callbacks.OnDeltaStreamClosed(streamID)
+		}
+	}()
+
+	// sends a response by serializing to protobuf Any
+	send := func(resp cache.DeltaResponse, typeURL string) (string, error) {
+		if resp == nil {
+			return "", errors.New("missing response")
+		}
+
+		out, err := resp.GetDeltaDiscoveryResponse()
+		if err != nil {
+			return "", err
+		}
+
+		// increment nonce
+		streamNonce = streamNonce + 1
+		out.Nonce = strconv.FormatInt(streamNonce, 10)
+		if s.callbacks != nil {
+			s.callbacks.OnStreamDeltaResponse(streamID, resp.GetDeltaRequest(), out)
+		}
+
+		return out.Nonce, str.Send(out)
+	}
+
+	// setState updates the currently known state of resources in the server
+	setState := func(res cache.DeltaResponse, typeURL string) {
+		values.mu.Lock()
+		defer values.mu.Unlock()
+
+		state := values.deltaStreamStates[typeURL]
+		if state.ResourceVersions == nil {
+			state.ResourceVersions = make(map[string]string)
+		}
```

Comment:
This doesn't seem necesssary since we overwrite `state.ResourceVersions` on the next line anyways?

Reply:
The reason I had that there is if we execute the next line without a nil `ResourceVersions` map, the code will panic.

Reply:
How so? On L184 we immediately overwrite the value we'd set on L182

Reply:
I ended up doing something like this:
```go
			var state stream.StreamState
			// Initialize the state map if we haven't already.
			// This means that we're handling the first request for this type on the stream.
			if s, ok := watches.deltaStreamStates[req.GetTypeUrl()]; !ok {
				state.ResourceVersions = make(map[string]string)
			} else {
				state = s
			}
```
But what I was originally seeing was go complaining about assignment to a nil map, I was similarly confused but seemed to get around it with adding that check